### PR TITLE
feat(testing): explicit resource management for `spy()` and `stub()`

### PR DESCRIPTION
**Pull request by KyleJune**
Closes https://github.com/denoland/deno_std/issues/3966

This is a work in progress. It won't work until [swc is upgraded to v1.4.0](https://github.com/denoland/deno_ast/issues/210) because earlier versions do not support the `using` keyword on functions. Until that issue is resolved, the new test cases I added fail with the following errors.

image

Once that issue is resolved, I believe the new test cases in this PR will pass and we could go forward with shipping it.

Regarding documentation, I added a description to both the spy and stub sections to explain that they are disposable. I noticed that FakeTime was already made disposible and it's example uses the `using` keyword. I thought it would be better to explicitly explain that since `using` is a new feature to TypeScript. I think if all the examples were updated to just use the `using` keyword to automate cleanup, people could easily miss the importance of that keyword in ensuring that the tests are cleaned up properly. Feel free to change my descriptions though.

### CODE REVIEW

File: testing/time.ts [hunk 1]
```diff
@@ -211,7 +211,7 @@ let dueTree: RedBlackTree<DueNode>;
  * });
  * ```
  */
-export class FakeTime {
+export class FakeTime implements Disposable {
```
Comment (KyleJune) [inline testing/time.ts:214]: This class was implementing Disposable already but didn't explicitly say that it was. I updated it to show that it implements Disposable to make it more clear that it is usable.

Comment (iuioiua) [reply testing/time.ts:214]: Good catch! Though, let's do this in a different PR. We could probably do PR that ensures that `[Async]Disposable` is implemented across the codebase where necessary.

**Comment by iuioiua** [comment]
Note to self: run CI when the required functionality is available in canary. Tests should pass in canary but fail in stable.

**Comment by iuioiua** [comment]
> Note to self: run CI when the required functionality is available in canary. Tests should pass in canary but fail in stable.

I didn't notice, but this is already the case! 😆 We've already upgraded to `swc@1.4.0`. I'll review this in time for the next release of `deno_std`.

**Comment by iuioiua** [comment]
@KyleJune, would you also like to update any applicable spies and stubs across the codebase to use the `using` keyword, too?

**Comment by babiabeo** [comment]
I think we should also add a note about the minimum Deno version required for it to work. 👍

**Comment by KyleJune** [comment]
> @KyleJune, would you also like to update any applicable spies and stubs across the codebase to use the `using` keyword, too?

Done, the only ones outside of the context of testing `testing/mock_test.ts` and `testing/time_test.ts` were `testing/bdd_test.ts` and `http/file_server_test.ts`. I didn't update the mock and time tests since the restore abilities are part of what is being tested.

> I think we should also add a note about the minimum Deno version required for it to work. 👍

I don't think that is necessary and that wasn't done when `FakeTime` was updated to have explicit resource management. I don't know if the next version will be v1.40.5 or v1.41.0 so I'm not going to update it to say which version is required. If the core members want to update it to specify the minimum version required for it, they can make the change.

**Review by kt3k — APPROVED** [review]
LGTM. Looks a nice feature!

**Review by iuioiua — APPROVED** [review]
This is awesome. Thank you! LGTM.

**Pull request merged**